### PR TITLE
feat: new sector culler for the v9 file format sector tree

### PR DESCRIPTION
**Pull request by larsmoa**

### CODE REVIEW

File: viewer/packages/utilities/src/three/isBox3OnPositiveSideOfPlane.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+/**
+ * Returns true if all corner point of the box provides is on the "positive side"
+ * of the plane provided (i.e. the corner point has a non-negative signed distance to the
+ * plane)
+ */
+export function isBox3OnPositiveSideOfPlane(box: THREE.Box3, plane: THREE.Plane): boolean {
+  let planeAccepts = false;
+  visitBox3CornerPoints(box, boundCorner => {
+    planeAccepts = plane.distanceToPoint(boundCorner) >= 0 || planeAccepts;
```
Comment (christjt) [inline viewer/packages/utilities/src/three/isBox3OnPositiveSideOfPlane.ts:15]: This is not the most efficient way to do this. It's fine I guess if you don't call this often. 

Better is to take the dot product betweem the normal of the plane and the vector from point on plane to the bound corner. If this is greater than 0, then the point is on the side of the plane which the normal is facing.

File: viewer/packages/cad-geometry-loaders/src/sector/culling/WeightFunctionsHelper.ts [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    // Project to NDC
+    p0.project(this._camera);
+    p1.project(this._camera);
+
+    // Factor saying how much the biggest node covers on screen (1 means full view)
+    const screenCoverage = p1.distanceToSquared(p0) / 4.0;
+
+    // Scale coverage such that 5% screen estate is 1.0 weight
+    const weight = Math.min(1.0, screenCoverage / 0.05);
+    return weight;
```
Comment (christjt) [inline viewer/packages/cad-geometry-loaders/src/sector/culling/WeightFunctionsHelper.ts:147]:
```suggestion
    return Math.min(1.0, screenCoverage / 0.05);
```

File: viewer/packages/cad-geometry-loaders/src/sector/culling/WeightFunctionsHelper.ts [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+  }
+
+  /**
+   * Computes a weight based on placement in sector tree (i.e. prioritize)
+   * sectors right below root sector.
+   * @param sector
+   */
+  computeSectorTreePlacementWeight(sector: SectorMetadata): number {
+    const levelWeight = sector.depth === 2 ? 1.0 : 1.0 / 3.0;
+    return levelWeight;
```
Comment (christjt) [inline viewer/packages/cad-geometry-loaders/src/sector/culling/WeightFunctionsHelper.ts:116]:
```suggestion
    return sector.depth === 2 ? 1.0 : 1.0 / 3.0;
```

**Review by christjt — CHANGES_REQUESTED** [review]

File: viewer/packages/cad-geometry-loaders/src/sector/culling/TakenSectorMap.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  initializeScene(modelMetadata: CadModelMetadata) {
+    this._takenSectorTrees.set(modelMetadata.modelIdentifier, {
+      sectorTree: new TakenSectorTree(modelMetadata.scene.root, this.determineSectorCost),
+      modelMetadata
+    });
+  }
+
+  getWantedSectorCount(): number {
+    let count = 0;
+    this._takenSectorTrees.forEach(({ sectorTree }) => {
```
[the remark this answers is not in the document]
Comment (larsmoa) [reply viewer/packages/cad-geometry-loaders/src/sector/culling/TakenSectorMap.ts]: No

File: examples/src/utils/CadBudgetUi.ts [hunk 5]
```diff
@@ -27,7 +27,7 @@ export function initialCadBudgetUi(viewer: Cognite3DViewer, uiFolder: dat.GUI) {
       maximumRenderCost: limitRenderCost ? defaultCadBudget.maximumRenderCost * scale : Infinity,
     };
   };
-  uiFolder.add(state, 'factor', 1, 500, 5).name('Budget factor (%)').onChange(onCadBudgetSettingsChanged);
+  uiFolder.add(state, 'factor', 1, 300, 1).name('Budget factor (%)').onChange(onCadBudgetSettingsChanged);
```
[the remark this answers is not in the document]
Comment (larsmoa) [reply examples/src/utils/CadBudgetUi.ts]: Changed back

File: viewer/packages/cad-geometry-loaders/src/sector/culling/ByScreenSizeSectorCuller.ts [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    const bounds = new THREE.Box3();
+    return sectors.filter(sector => {
+      bounds.copy(sector.bounds);
+      bounds.applyMatrix4(modelMatrix);
+
+      const shouldKeep = clippingPlanes.every(plane => isBox3OnPositiveSideOfPlane(bounds, plane));
+      return shouldKeep;
+    });
+  }
+  return sectors;
```
Comment (christjt) [inline viewer/packages/cad-geometry-loaders/src/sector/culling/ByScreenSizeSectorCuller.ts:182]: I invert the if in cases like this to reduce branching:
```suggestion
  if (clippingPlanes.length <= 0) {
    return sectors;
  }
  
  const bounds = new THREE.Box3();
 
  return sectors.filter(sector => {
    bounds.copy(sector.bounds);
    bounds.applyMatrix4(modelMatrix);

    const shouldKeep = clippingPlanes.every(plane => isBox3OnPositiveSideOfPlane(bounds, plane));
    return shouldKeep;
  });
  
  
```

File: viewer/packages/cad-geometry-loaders/src/sector/culling/WeightFunctionsHelper.ts [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+      { near: near + 0.05 * nearFarRange, far: near + 0.4 * nearFarRange, weight: 0.7 }, // 5-40% of frustum
+      { near: near + 0.4 * nearFarRange, far: near + 1.0 * nearFarRange, weight: 0.2 } // 40-100% of frustum
+    ].map(x => {
+      const projectionMatrix = createModifiedProjectionMatrix(camera, x.near, x.far);
+      const frustumMatrix = new THREE.Matrix4().multiplyMatrices(projectionMatrix, this._camera.matrixWorldInverse);
+      const frustum = new THREE.Frustum().setFromProjectionMatrix(frustumMatrix);
+      return {
+        ...x,
+        frustum
+      };
```
Comment (christjt) [inline viewer/packages/cad-geometry-loaders/src/sector/culling/WeightFunctionsHelper.ts:47]:
```suggestion
    ].map(modifiedFrustum => {
      const projectionMatrix = createModifiedProjectionMatrix(camera, modifiedFrustum.near, modifiedFrustum.far);
      const frustumMatrix = new THREE.Matrix4().multiplyMatrices(projectionMatrix, this._camera.matrixWorldInverse);
      const frustum = new THREE.Frustum().setFromProjectionMatrix(frustumMatrix);
      return {
        ...modifiedFrustum,
        frustum
      };
```

File: viewer/packages/cad-geometry-loaders/src/sector/culling/WeightFunctionsHelper.ts [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    });
+  }
+
+  addCandidateSectors(sectors: SectorMetadata[], modelMatrix: THREE.Matrix4) {
+    // Note! We compute distance to camera, not screen (which would probably better)
```
Comment (christjt) [inline viewer/packages/cad-geometry-loaders/src/sector/culling/WeightFunctionsHelper.ts:52]:
```suggestion
    // Note! We compute distance to camera, not screen (which would probably be better)
```

File: viewer/packages/cad-geometry-loaders/src/sector/culling/WeightFunctionsHelper.ts [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    return frustumWeight;
+  }
+
+  /**
+   * Computes a weight based on placement in sector tree (i.e. prioritize)
+   * sectors right below root sector.
+   * @param sector
+   */
+  computeSectorTreePlacementWeight(sector: SectorMetadata): number {
+    return sector.depth === 2 ? 1.0 : 1.0 / 3.0;
```
Comment (christjt) [inline viewer/packages/cad-geometry-loaders/src/sector/culling/WeightFunctionsHelper.ts:115]: This seems kinda brittle and is VERY tied to the tree. I assume any hierarchical change here would totally brake this. If the hierarchy is set in stone then ok, but at least add a comment here (why exactly does depth 2 get a higher weight?).

File: viewer/packages/cad-geometry-loaders/src/sector/culling/takensectors/TakenSectorMapBase.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    const wanted = this.collectWantedSectors();
+    const nonDiscarded = wanted.filter(sector => sector.levelOfDetail !== LevelOfDetail.Discarded);
+
+    const totalSectorCount = modelsMetadata.reduce((sum, x) => sum + x.scene.sectorCount, 0);
+    const takenSectorCount = nonDiscarded.length;
+    const takenSimpleCount = nonDiscarded.filter(x => x.levelOfDetail === LevelOfDetail.Simple).length;
+    const forcedDetailedSectorCount = nonDiscarded.filter(x => !Number.isFinite(x.priority)).length;
+    const accumulatedPriority = nonDiscarded
+      .filter(x => Number.isFinite(x.priority) && x.priority > 0)
+      .reduce((sum, x) => sum + x.priority, 0);
```
Comment (christjt) [inline viewer/packages/cad-geometry-loaders/src/sector/culling/takensectors/TakenSectorMapBase.ts:23]: I prefer when the lambda functions have descriptive variable names, but I'll let you make that choice.

File: viewer/packages/cad-geometry-loaders/src/sector/culling/takensectors/TakenV9SectorMap.ts [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+  private readonly determineSectorCost: DetermineSectorCostDelegate<V9SectorMetadata>;
+  private readonly _totalCost: SectorCost = { downloadSize: 0, drawCalls: 0, renderCost: 0 };
+  private readonly _models = new Map<string, { modelMetadata: CadModelMetadata; sectorIds: Map<number, number> }>();
+
+  get totalCost(): SectorCost {
+    return { ...this._totalCost };
+  }
+
+  get modelsMetadata(): CadModelMetadata[] {
+    return Array.from(this._models.values()).map(x => x.modelMetadata);
```
Comment (christjt) [inline viewer/packages/cad-geometry-loaders/src/sector/culling/takensectors/TakenV9SectorMap.ts:23]:
```suggestion
    return Array.from(this._models.values()).map(cadModel => cadModel.modelMetadata);
```

File: viewer/packages/cad-geometry-loaders/src/sector/culling/takensectors/TakenV9SectorMap.ts [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+      const sectorMetadata = model.scene.getSectorById(sectorId);
+      assert(sectorMetadata !== undefined);
+
+      const sectorCost = this.determineSectorCost(sectorMetadata! as V9SectorMetadata, LevelOfDetail.Detailed);
+      addSectorCost(this._totalCost, sectorCost);
+
+      sectorIds.set(sectorId, priority);
+    } else {
+      sectorIds.set(sectorId, Math.max(priority, existingPriority));
+    }
```
Comment (christjt) [inline viewer/packages/cad-geometry-loaders/src/sector/culling/takensectors/TakenV9SectorMap.ts:59]: I would invert and return to reduce branching, but up to you.

**Review by christjt — APPROVED** [review]
Looks very nice and clean, a few nits / comments that you may or may not do.

File: viewer/packages/cad-geometry-loaders/src/sector/culling/WeightFunctionsHelper.ts (hunk 9 above)
Comment (larsmoa) [reply viewer/packages/cad-geometry-loaders/src/sector/culling/WeightFunctionsHelper.ts:115]: Added comment

**Pull request merged**